server/data/data-cassandra/src/test/java/org/apache/james/rrt/cassandra/migration/MappingsSourcesMigrationTest.java

```diff
@@ -0,0 +1,130 @@
+        cassandraRecipientRewriteTableDAO = new CassandraRecipientRewriteTableDAO(cassandra.getConf(), CassandraUtils.WITH_DEFAULT_CONFIGURATION);
+        cassandraMappingsSourcesDAO = new CassandraMappingsSourcesDAO(cassandra.getConf());
+
+        migration = new MappingsSourcesMigration(cassandraRecipientRewriteTableDAO, cassandraMappingsSourcesDAO);
+    }
+
+    @Test
+    void emptyMigrationShouldSucceed() {
+        assertThat(migration.run()).isEqualTo(Migration.Result.COMPLETED);
+    }
+
+    @Test
+    void migrationShouldSucceedWithData() {
+        cassandraRecipientRewriteTableDAO.addMapping(SOURCE, MAPPING);
+
+        assertThat(migration.run()).isEqualTo(Task.Result.COMPLETED);
+    }
+
+    @Test
+    void migrationShouldCreateMappingSourceFromMapping() {
+        cassandraRecipientRewriteTableDAO.addMapping(SOURCE, MAPPING);
+
+        migration.run();
+
+        assertThat(cassandraMappingsSourcesDAO.retrieveSources(MAPPING).collectList().block())
+            .containsExactly(SOURCE);
+    }
+
+    @Test
+    void migrationShouldCreateMultipleMappingSourcesFromMappings() {
+        MappingSource source2 = MappingSource.fromUser("bob", Domain.LOCALHOST);
+
+        cassandraRecipientRewriteTableDAO.addMapping(SOURCE, MAPPING);
+        cassandraRecipientRewriteTableDAO.addMapping(source2, MAPPING);
+
+        migration.run();
+
+        assertThat(cassandraMappingsSourcesDAO.retrieveSources(MAPPING).collectList().block())
+            .containsOnly(SOURCE, source2);
+    }
+
+    @Test
+    void migrationShouldReturnPartialWhenGetAllMappingsFromMappingsFail() {
+        CassandraRecipientRewriteTableDAO cassandraRecipientRewriteTableDAO = mock(CassandraRecipientRewriteTableDAO.class);
+        CassandraMappingsSourcesDAO cassandraMappingsSourcesDAO = mock(CassandraMappingsSourcesDAO.class);
+        migration = new MappingsSourcesMigration(cassandraRecipientRewriteTableDAO, cassandraMappingsSourcesDAO);
+
+        when(cassandraRecipientRewriteTableDAO.getAllMappings()).thenThrow(new RuntimeException());
+
+        assertThat(migration.run()).isEqualTo(Migration.Result.PARTIAL);
+    }
+
+    @Test
+    void migrationShouldReturnPartialAddMappingFails() {
+        CassandraRecipientRewriteTableDAO cassandraRecipientRewriteTableDAO = mock(CassandraRecipientRewriteTableDAO.class);
+        CassandraMappingsSourcesDAO cassandraMappingsSourcesDAO = mock(CassandraMappingsSourcesDAO.class);
+        migration = new MappingsSourcesMigration(cassandraRecipientRewriteTableDAO, cassandraMappingsSourcesDAO);
+
+        when(cassandraRecipientRewriteTableDAO.getAllMappings())
+            .thenReturn(Mono.just(ImmutableMap.of(SOURCE, MappingsImpl.fromMappings(MAPPING))));
+        when(cassandraMappingsSourcesDAO.addMapping(any(Mapping.class), any(MappingSource.class)))
+            .thenThrow(new RuntimeException());
+
+        assertThat(migration.run()).isEqualTo(Migration.Result.PARTIAL);
+    }
+
+
```

Comment:
What happens if I run the migration 2 times?

I should not get the error, and listSourcesDAO should still contain only the right entries.

Reply:
Could we start writing concurrent tests for migrations? (10 threads + 10 entries)

Have a look to `ConcurrentTestRunner`